[SPARK-50343][SPARK-50344][SQL] Add SQL pipe syntax for the DROP and AS operators

dtenedor opened the pull request:
### What changes were proposed in this pull request?

This PR adds SQL pipe syntax support for DROP and AS operators.

The DROP operator removes one or more existing column from the input table.
The AS operator assigns a new table alias to the preceding relation.

These are equivalent to `SELECT * EXCEPT(<columnNames>)` and `SELECT ... FROM (...) <tableAlias>`, respectively.

For example:

```
values (0, 'pqr', 2), (3, 'tuv', 5) as tab(a, b, c)
|> as v
|> drop v.b
|> drop v.c

0
3
```

This PR also fixes a small bug and adds more testing.

### Why are the changes needed?

The SQL pipe operator syntax will let users compose queries in a more flexible fashion.

### Does this PR introduce _any_ user-facing change?

Yes, see above.

### How was this patch tested?

This PR adds a few unit test cases, but mostly relies on golden file test coverage. I did this to make sure the answers are correct as this feature is implemented and also so we can look at the analyzer output plans to ensure they look right as well.

### Was this patch authored or co-authored using generative AI tooling?

No

dtenedor committed:
update

update

update

update

update

update

update

dtenedor committed:
add q3

dtenedor committed:
add q12

dtenedor committed:
add q12

dtenedor committed:
add q44

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -5106,7 +5106,7 @@ object SQLConf {
       .version("4.0.0")
       .booleanConf
-      .createWithDefault(Utils.isTesting)
+      .createWithDefault(true)
```
Is this a mistake? This should be a part of this PR, @dtenedor .

dtenedor replied:
@dongjoon-hyun Good question; this is intentional, the feature is done as of this PR and ready to be enabled by default. There are no backwards compatibility issues as the new syntax will work with traditional SQL queries with no breaking changes.

Let me know if you prefer to split this config into a separate PR given this information and I can do that.

dtenedor commented:
cc @dongjoon-hyun @cloud-fan @gengliangwang this is ready for review at your convenience.

dongjoon-hyun reviewed (dismissed):
To @dtenedor , +1 for your passion. 😄 
> Good question; this is intentional

However, all open-source communities including Apache Spark community prefer not to mix independent themes into a single PR. Especially, the PR title, `Add SQL pipe syntax for the DROP and AS operators` doesn't cover your silent piggy-backing. 

> Let me know if you prefer to split this config into a separate PR given this information and I can do that.

Yes, to simply put, please proceed it separately to get a proper review.

dtenedor committed:
respond to code review comments

dtenedor commented:
@dongjoon-hyun sounds good, I switched the config change back for this PR and can enable it later in a subsequent PR instead.

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -6013,6 +6011,18 @@ class AstBuilder extends DataTypeAstBuilder
+      Project(projectList, left)
+    }.getOrElse(Option(ctx.AS).map { _ =>
+      val child = left match {
+        case s: SubqueryAlias => s.child
```
shall we keep it as it is in analyzer? It will be eliminated later anyway.

gengliangwang commented on sql/core/src/test/resources/sql-tests/inputs/pipe-operators.sql:
```diff
@@ -320,6 +770,84 @@ table t
+
+-- Multiple aliases are not supported.
+table t
+|> as u, v;
```
what about 
```
as `u.v`
```

gengliangwang replied:
Also, let's test simple invalid cases such as `a-b` too

gengliangwang commented on sql/core/src/test/resources/sql-tests/inputs/pipe-operators.sql:
```diff
@@ -71,6 +71,456 @@ create temporary view windowTestData as select * from values
   (3, 1L, 1.0D, date("2017-08-01"), timestamp_seconds(1501545600), null)
   AS testData(val, val_long, val_double, val_date, val_timestamp, cate);
 
+create table store_sales(
```
Shall we move the TPCDS table creation/drop to the `SQLQueryTestSuite`, so that we can review the outputs easier.
We can have something like `def requireTPCDSCases: Seq[String]` to define which golden file needs the TPCDS table definitions.

dtenedor replied:
Sure, this is done. I found an existing trait `TPCDSSchema` that can help with this.

dtenedor committed:
respond to code review comments

respond to code review comments

respond to code review comments

respond to code review comments

dtenedor reviewed (commented):
Thank you @gengliangwang for your review!

dongjoon-hyun reviewed (approved):
+1, LGTM. Thank you, @dtenedor and @gengliangwang .

Also, cc @yaooqinn , too

dongjoon-hyun commented:
BTW, I added `SPARK-50344` to the PR title because this covers `AS` too, @dtenedor .

gengliangwang commented:
Thanks, merging to master
